proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,710 @@
+creating nested instances. In general, with this proposal,
+`module_instantiate(M)` goes from always creating a single instance to
+potentially creating a *DAG* of instances, with `M`'s instance as the root. 
+
+
+### Module and Instance References
+
+The instruction `ref.instance` used in the previous section has the signature:
+```wasm
+ref.instance $instance : [] -> [ (ref $InstanceT) ]
+```
+where `$instance` refers to an instance (in the instance index space) of type
+`$InstanceT`. Types of the form `(ref (instance ...))` are a new kind of
+*instance reference type*.
+
+The proposal also adds a symmetric instruction for modules:
+```wasm
+ref.module $module : [] -> [ (ref $ModuleT) ]
+```
+where `$module` refers to a module (in the module index space) of type
+`$ModuleT`. Types of the form `(ref (module ...))` are a new kind of
+*module reference type*. (See the
+[shared-everything dynamic linking example](Example-SharedEverythingDynamicLinking.md)
+to see sample usage.)
+
+Similar to the subtype relationship that exists between typed function
+references and `funcref` in the [Function References] proposal, module reference
+types share a supertype `moduleref` and instance reference types share a
+supertype `instanceref`:
+```wasm
+(ref (func ...)) <: funcref
+(ref (module ...)) <: moduleref
+(ref (instance ...)) <: instanceref
+```
+
+The previous section noted that `instance.instantiate` is, in this proposal,
+only allowed in the constant initializer expressions of `instance` definitions.
+The immediate reason for this is the GC requirement stated [above](#additional-requirements):
+since instances can form graphs (via mutable tables and globals), a first-class
+`instance.instantiate` instruction would inherently require a garbage collection
+algorithm to collect dead instances. This could be added later on as part of a
+batch of GC features (viz., [GC types]) with a careful definition of what
+features are available on hosts that don't support GC. But for now, this proposal
+aims to be universally available.
+
+In contrast, `ref.module`/`ref.instance` and module/instance references have no
+such inherent GC requirement and would be useful in allowing hosts to optionally
+expose runtime instantiation through *import APIs*. For example, using
+[Interface Types] to express the signature more succinctly, a classic
+[`dlopen`]/[`dlsym`] interface could be approximated:
+```wasm
+(instance
+  (export "dlopen" (func
+    (param $module moduleref)
+    (param $imports (array (record (field "name" string) (field "value" anyref))))
+    (result instanceref)
+  ))
+  (export "dlsym" (func
+    (param $instance instanceref)
+    (param $symbol string)
+    (result funcref)
+  ))
+)
+```
+
+Thus, as a conservative first step toward runtime dynamic linking, this proposal
+includes `ref.module` and `ref.instance` in the general wasm instruction set and
+includes module and instance reference types in the general [`reftype`] set.
+
+
+### Nested Modules
```

Comment:
I think nested modules are an essential part of this proposal.  One basic reason is that for every definable kind of thing T in wasm, we have T imports, T definitions and T exports.  Since we have module imports, it would then follow we have module definitions (i.e., nested modules).  But also, practically, speaking, nested modules are useful in their own right as shown in the examples in this proposal.

For your `(instance $i (func $hi ...))` example, are you saying that's syntactic sugar for:
```
(module $I (func (export "hi") ...))
(instance $i (instance.instantiate $I))
```
?  I suppose it's worth considering, but I wonder how useful it is if there are no imports.

Reply:
I added a sentence "In the future, new productions could be added to <instance-init> allowing alternatives to instantiate for creating instances, such as directly from fields without an intermediate module." to the explainer that mentions this kind of future extension.